tree/ntuple/v7/test/ntuple_descriptor.cxx

```diff
@@ -123,3 +123,99 @@ TEST(RNTuple, Descriptor)
+       ntuple.Fill();
+   }
+
+   RNTupleReader ntuple(std::move(modelRead),
+      std::make_unique<RPageSourceFile>("ntuple", fileGuard.GetPath(), RNTupleReadOptions()));
+
+   // iterate over top-level fields
+   std::vector<std::string> names{};
+   for (auto& f: ntuple.GetDescriptor().GetTopLevelFields()) {
+      names.push_back(f.GetFieldName());
+   }
+   EXPECT_EQ(names.size(), 5);
+   EXPECT_EQ(names[0], std::string("jets"));
+   EXPECT_EQ(names[1], std::string("jets"));
+   EXPECT_EQ(names[2], std::string("bools"));
+   EXPECT_EQ(names[3], std::string("bool_vec_vec"));
+   EXPECT_EQ(names[4], std::string("ints"));
+
+   const auto& ntuple_desc = ntuple.GetDescriptor();
+   auto top_level_fields = ntuple_desc.GetTopLevelFields();
+
+   // iterate over child field ranges
+   const auto& float_vec_desc = *top_level_fields.begin();
+   EXPECT_EQ(float_vec_desc.GetFieldName(), std::string("jets"));
+   auto float_vec_child_range = ntuple_desc.GetFieldRange(float_vec_desc);
+   std::vector<std::string> child_names{};
+   for (auto& child_field: float_vec_child_range) {
+      child_names.push_back(child_field.GetFieldName());
+      // check the empty range
+      auto float_child_range = ntuple_desc.GetFieldRange(child_field);
+      EXPECT_FALSE(float_child_range.begin() != float_child_range.end());
+   }
+   EXPECT_EQ(child_names.size(), 1);
+   EXPECT_EQ(child_names[0], std::string("float"));
+
+   // check if canonical iterator methods work
+   auto iter = top_level_fields.begin();
+   std::advance(iter, 3);
+   const auto& bool_vec_vec_desc = *iter;
+   EXPECT_EQ(bool_vec_vec_desc.GetFieldName(), std::string("bool_vec_vec"));
+
+   child_names.clear();
+   for (auto& child_field: ntuple_desc.GetFieldRange(bool_vec_vec_desc)) {
+      child_names.push_back(child_field.GetFieldName());
+   }
+   EXPECT_EQ(child_names.size(), 1);
+   EXPECT_EQ(child_names[0], std::string("std::vector<bool>"));
+}
+
+TEST(RFieldDescriptorRange, SortByFieldNames)
+{
+   auto model = RNTupleModel::Create();
+   auto floats = model->MakeField<std::vector<float>>("jets");
+   auto floats2 = model->MakeField<std::vector<float>>("jets");
+   auto bools = model->MakeField<std::vector<bool>>("bools");
+   auto bool_vec_vec = model->MakeField<std::vector<std::vector<bool>>>("bool_vec_vec");
+   auto ints = model->MakeField<std::int32_t>("ints");
+
+   FileRaii fileGuard("test_field_iterator.root");
+   auto modelRead = std::unique_ptr<RNTupleModel>(model->Clone());
+   {
+       RNTupleWriter ntuple(std::move(model),
+          std::make_unique<RPageSinkFile>("ntuple", fileGuard.GetPath(), RNTupleWriteOptions()));
+       ntuple.Fill();
+   }
+
+   RNTupleReader ntuple(std::move(modelRead),
+      std::make_unique<RPageSourceFile>("ntuple", fileGuard.GetPath(), RNTupleReadOptions()));
+
+   std::vector<std::string> sorted_names{};
+   for (auto& f: ntuple.GetDescriptor().GetTopLevelFields().SortByNames()) {
```

Comment:
On second glance, this is possibly UB (see Temporary range expression section of https://en.cppreference.com/w/cpp/language/range-for)

Reply:
I am now reasonably sure this is not UB, since `ntuple` is always a valid reference, and a simplified example on Compiler Explorer was clean (https://godbolt.org/z/nvkvKL).  I think the warning on cppreference would apply if `ntuple` was also a temporary, i.e. 
```cpp
for (auto& f: GetNTuple().GetDescriptor().GetTopLevelFields().SortByNames())
```

Reply:
I believe this is UB as `GetTopLevelFields()` returns a temporary; this is not lifetime-extended. Only the temporary returned by `SortByNames()` is lifetime-extended. See e.g. https://abseil.io/tips/107 on how lifetime-extension works, and why it works for the final expression.

The UB sanitizer cannot check for all UB - though this would indeed be a good thing to have diagnosed.

Given that we have at least some concern about this being UB: could you simply assign `ntuple.GetDescriptor().GetTopLevelFields()` to a local variable before the `for`-loop?

Reply:
Since .SortByNames() returns by value is that really a problem that the return value of GetTopLevelFields() is destructed?

Reply:
Do you want the test to survive a change in the implementation of the classes invilved? I would say that's a definitive plus :-) I.e. it really depends on how exactly these objects look like - whether they have any references between them or not, and that might well change at some point in the future.

I believe an extra variable is a fair price for avoiding UB, guaranteed.

Reply:
Adding a local variable solves the problem here, but this is still a footgun for user code (the test demonstrates the public API of the NTuple). Maybe there is an alternative implementation that works? (e.g. mutating the class in place)

Reply:
I believe we are able to sidestep this issue altogether with the new sorting method in e05a85b37988b